### PR TITLE
PS-5333: Remove need for innodb_encrypt_tables to ON/FORCE when

### PR DESCRIPTION
**Pull request by robgolebiowski**
mysql.ibd is encrypted

PS-5117 introduced requirement that innodb_encrypt_tables need to be set
to ON/FORCE on server startup when mysql.ibd is encrypted. This
requirement was added because there was a need for mysql.ibd space flags
validation against hardcoded, predefined flags. In PS mysql.ibd can be
encrypted and have encrypted flag set - thus generating a mismatch with
predefined flags. In order to check whether mysql.ibd is encrypted we
will check flags from the first page of the tablespace instead of
relaying on innodb_encrypt_tables. Apart from improvement from user
point of view it will also ease keyring encryption threads development
in 8.0 as encryption threads can also change the mysql.ibd's encryption
flag.

In keyring encryption there is a possibility that update to mysql.ibd's
page0 will reside only inside redo log (due to the crash).
Added code for reading page0 encryption flag from redo log and using it
to update mysql.ibd's DD flag. This code will be tested when keyring
encryption gets fixed.

### CODE REVIEW

**Comment by robgolebiowski** [comment]
This PR is exactly the same as https://github.com/percona/percona-server/pull/2937. Here, on top of 2937, I also implemented checks for mysql.ibd's page 0 updates in redo log suggested by Satya. This code is not currently tested as there is no scenario in MK encryption where there are updates to encryption flag stored in page0 of mysql.ibd. However, this code is needed for keyring encryption and testes for this code will be added later, when keyring encryption test suite will be able to be run in jenkins.

File: storage/innobase/dict/dict0dict.cc
Comment (laurynas-biveinis) [inline storage/innobase/dict/dict0dict.cc:7514]: suspicious: pfs_ types should not be used directly: os_file_t ?

File: storage/innobase/dict/dict0dict.cc
Comment (laurynas-biveinis) [inline storage/innobase/dict/dict0dict.cc]: This looks useful enough to have it in e.g. ut0ut.h: unique_ptr with ut_free deleter (but with struct ut_free_deleter w/ operator() instead of std::function<>/lambda to avoid having state in the smart pointer)

File: storage/innobase/srv/srv0start.cc
Comment (laurynas-biveinis) [inline storage/innobase/srv/srv0start.cc]:
```suggestion
      // We have set mysql's dd flags based on what we read from page 0. This was
```

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7514]: Seems that pfs_os_file_t is used everywhere where os_file_create_simple_no_error_handling is used ...

Comment (laurynas-biveinis) [reply storage/innobase/dict/dict0dict.cc:7514]: Huh. OK then

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc]: added to ut0new.h

**Review by laurynas-biveinis — APPROVED** [review]

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc]: Please keep the function names precise. Exact details of function can be added as function comment.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:7560]: Please add doxygen style comments for all new functions

File: storage/innobase/srv/srv0start.cc
Comment (satya-bodapati) [inline storage/innobase/srv/srv0start.cc:2705]: you are not fixing the fil_space_t flags of mysql tablespace. Only fixing DD is not sufficient.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc]: I disagree. Comments then need to change when function changes. I took this from clean code and I will stand by it. Self explanatory names are much better - imho.

File: storage/innobase/srv/srv0start.cc
Comment (robgolebiowski) [reply storage/innobase/srv/srv0start.cc:2705]: mysql tablespace fil_space_t flags was already updated by redo log applier.

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc]: I disagree. Function names should describe what the function is doing. Comments are additional maintenance burden and thus should be avoided if possible. Thus the code should be self explanatory. I have changed the names to be more generic and describe what the function is doing instead of describing what it is used for.

File: storage/innobase/srv/srv0start.cc
Comment (robgolebiowski) [reply storage/innobase/srv/srv0start.cc:2705]: mysql tablespace fil_space_t flags was already updated by redo log applier.

File: storage/innobase/handler/ha_innodb.cc
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]: Why do you need to pass a global variable?

File: storage/innobase/handler/ha_innodb.cc
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]: You are not supposed to re-use ER_ messages. Every message should introduce a unique id.

File: storage/innobase/handler/ha_innodb.cc
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]: Did you mean "have you deleted it?"

File: sql/dd/types/init_mode.h
Comment (satya-bodapati) [inline sql/dd/types/init_mode.h:3]: Why Percona copyright? This is just Oracle code in different file

File: storage/innobase/include/dict0dd.h
Comment (satya-bodapati) [inline storage/innobase/include/dict0dd.h]: Shouldn't this be static function?

File: storage/innobase/include/dict0dd.h
Comment (satya-bodapati) [inline storage/innobase/include/dict0dd.h:1235]: Remove this function. Dead code

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc:6133]: Remove this. Dead code

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc:6172]: With the clear and set functions unused, only dd_set_flags is remaining and there is no real reason to use std::function.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc]: I see that you really like long names (function name that can go upto 60-80 characters).

Currently this handles only  ON or FORCE. So you made a choice to a "self describing" function which mentions value as part of function name. So if this function extends to handle below values,
SRV_ENCRYPT_TABLES_KEYRING_ON:
SRV_ENCRYPT_TABLES_KEYRING_FORCE:
SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING:
SRV_ENCRYPT_TABLES_ONLINE_TO_KEYRING_FORCE:

the function name should be:
dict_is_encrypt_tables_either_on_or_force_or_keyring_on_or_keyring_force_or_online_keyring_to_keyring_or_online_to_keyring_force() ? isn't it ;)

IMHO, function name should be precise and clear. This doesn't mean "Long explaining function names".

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc:6175]: General suggestion "if_needed" can be avoided most times.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:7514]: You should avoid direct reads if possible. (may be buf_page_get_gen()?)

Is the double write buffer recovery over at this stage? Does the double write buffer do "sync" io on the tablespace? -> If the answer is yes to both of them, then you may use direct read.

File: storage/innobase/include/ut0new.h
Comment (satya-bodapati) [inline storage/innobase/include/ut0new.h:1223]: Only this variant is required for now. Rest is dead code.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc]: Ok, I disagree too.

Function names should be precise. It doesn't need to describe the entire functionality in "name". That is why we have function comments. 

Fixing function comments when function does extra is not "burden". Actually it is more burden when you function names like this. Let say function does more. So we have to fix function name  fomr "it_does_this()" to "it_does_this_and_more()".

So you have to fix all the callers as well.

See my comment in the place where you use "is_on_or_force()"

File: storage/innobase/srv/srv0start.cc
Comment (satya-bodapati) [reply storage/innobase/srv/srv0start.cc:2705]: Ok, is it possible to add assertion that flags have encrypted bit set?

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc]: I think it was agreed that we are not code reviewing individual codying styles. I opt for longer names describing what a function is doing. You can disagree, but you cannot hold PR because of that.

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7560]: Doxygen comments were added to header file in this case. I do not think comments repetition makes sense - just add maintenance cost.

File: storage/innobase/srv/srv0start.cc
Comment (robgolebiowski) [reply storage/innobase/srv/srv0start.cc:2705]: Why encrypted bit set ? it might be that the bit is not set. It's checking if mysql table was changed by redo applier and then it sets encryption flag in DD to the one from fil_space_t.

File: sql/dd/types/init_mode.h
Comment (robgolebiowski) [reply sql/dd/types/init_mode.h:3]: But the file is ours, so there was a change. Laurynas asked me to put Oracle copyrights here, next to Percona one.

File: storage/innobase/dict/dict0dd.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dd.cc:6172]: clear and set functions are used by the encryption threads. I have taken the DD flag update code from there. Turns out that your comment about fixing DD encryption flag for mysql.ibd is only valid for encryption threads. I do not want to put effort in making code changes that will need to be reverted once encryption threads code is merged in. There I will also test the mysql.ibd DD code - so this PR depends on encryption threads anyways.

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc]: I disagree, but this is code style and I will stay with longer names.

File: storage/innobase/dict/dict0dd.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dd.cc:6175]: Well it gives you information that mysql_ibd will not always need flag fixing - without studying the body of dd_fix_mysql_ibd_encryption_flag

File: sql/dd/types/init_mode.h
Comment (satya-bodapati) [reply sql/dd/types/init_mode.h:3]: @laurynas-biveinis Are you sure with Percona copyright here? There is nothing new by percona here.

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dd.cc:6175]: ok, similar discussion in other thread. Many function has "if" statements. So this means the decision "if needed" can be written to all.  So functions
"dict_set_flags_if_needed()",  "dict_update_information_if_val_is_true()"
I know  this is 'your' preferred way. Had to say anyway.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:7560]: If it is in header, it is fine.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc]: Ok. Enjoy using them.

Other than that, why are you handling only ON and FORCE? what about the other enum values?

File: storage/innobase/srv/srv0start.cc
Comment (satya-bodapati) [reply storage/innobase/srv/srv0start.cc:2705]: Ok, I see

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc]: Anyways, you can keep this as you wish.

Am not holding this PR for this reason. There are other comments, concerns raised. If those are solved, it can proceed further.

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dd.cc:6172]: Ok, Do you have testcase for exercising all this code (by encryption threads)?
Should be easy to write one (not as part of this PR )
disable flush, encryption thread encrypt mysql.ibd, crash and ensure startup works fine.

File: sql/dd/types/init_mode.h
Comment (laurynas-biveinis) [reply sql/dd/types/init_mode.h:3]: Our new files should carry our copyright. Let's keep both copyrights and let someone better versed in these matters to change it later if needed

Comment (satya-bodapati) [reply sql/dd/types/init_mode.h:3]: Ok

File: storage/innobase/dict/dict0dd.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dd.cc:6172]: yes, this will be added in encryption threads test suite.

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7514]: I have implemented reading from the buffer, as you said, but it is not possible to do so at this stage - because the fil_system is not yet ready. When I tried to acquire the tablespace I got missing tablespace error and when I tried to read page from a buffer without acquiring space I got a core dump - as assert that page's space must exists in the fil_system was false.

File: storage/innobase/include/ut0new.h
Comment (robgolebiowski) [reply storage/innobase/include/ut0new.h:1223]: @laurynas-biveinis wanted to see those functions for future use

File: storage/innobase/include/dict0dd.h
Comment (robgolebiowski) [reply storage/innobase/include/dict0dd.h:1235]: is used in keyring encryption. Will be used when it gets merged in.

File: storage/innobase/include/dict0dd.h
Comment (robgolebiowski) [reply storage/innobase/include/dict0dd.h]: is used in keyring encryption. Will be used when it gets merged in.

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dd.cc:6172]: Does it exist in your encryption branch?

Comment (robgolebiowski) [reply storage/innobase/dict/dict0dd.cc:6172]: one of the tests (more) was failing due to this. I have restarted the server after all tables were encrypted with encryption threads and I got a crash as the DD was not expecting mysql.ibd to be encrypted.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:7514]: I have done lot of debugging and code study and I now am in better position :)

In my debugging, I found that recovery (recv_recovery_from_checkpoint_finish()) is over before dict_detect_encryption_of_mysql_ibd().

So there shouldn't be any issues.

1. Why is fil_system not ready. If redo recovery is over, then fil_system is fully initialized.
2. Why is mysql.ibd tablespace missing in fil_system cache?
3. buf_page_get_gen() should work fine at this stage. Infact recovery also uses buffer pool read to apply redo on pages. See buf_read_recv_pages().

And to answer the questions I asked.
"Is the double write buffer recovery over at this stage? 
Yes, the dblwr buffer recovery is over before even redo starts applying/scanning.

Does the double write buffer do "sync" io on the tablespace?
Yes, The IO is sync IO and also fsync() is done.

Neverthless, based on my debugging and observations, it is not required to do direct IO. A simple buffer pool read should work.

Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:7514]: Ok, after our discussion our on slack and then verifying with the branch, there is *indeed* a case where we cannot do buffer pool read. This is the non-recovery case. But there is still one issue.

If recovery fixed mysq.ibd, all the changes to page 0 of mysql are done via buffer pool and regular IO.

If you do direct IO in this case, you may read a stale copy of page. The updated page 0 copy is in buffer pool (and IO threads may write at leisure).

File: storage/innobase/srv/srv0start.cc
Comment (satya-bodapati) [inline storage/innobase/srv/srv0start.cc:2704]: if the tablespace is opened by redo, it should not be nullptr.
Can you add ut_ad(mysql_space != nullptr)?

File: storage/innobase/dict/dict0dict.cc
Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7514]: @satya-bodapati I will then implement first reading from buffer pool and then if the page does not exist I will read with regular IO

Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:7514]: Yes, May be you should use  is_mysql_ibd_page_0_in_redo() (save state). If this is true, then fil_space_t* should be present and you should do buffer pool read, else { do direct IO read)

**Comment by robgolebiowski** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-param/72/

File: storage/innobase/dict/dict0dd.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dd.cc:6190]: Can you please add ib::error() here with the flags values? Also mention this is with encryption flag excluded.

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:7535]: Can you please format into doxygen style used by other functions? Only this one seems a bit different

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc]: s/disc/ disk

**Review by satya-bodapati — APPROVED** [review]

File: storage/innobase/dict/dict0dict.cc
Comment (satya-bodapati) [inline storage/innobase/dict/dict0dict.cc:7504]: block should never be nullptr. another debug assert here please

Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7504]: actually with PS implementation it can be nullptr. If srv_pass_corrupt_tables is ON or there is a decryption error. For MySQL it would just crash.

Comment (satya-bodapati) [reply storage/innobase/dict/dict0dict.cc:7504]: Ok, does it mean all buf_page_get_gen() in PS checks for nullptr?

Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7504]: I cannot say for sure all of them, but we should check for null when we add new code. In encryption threads branch I have added quite a lot of them - to see if decryption passed.

Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7504]: sorry, when decryption fails I am returning error though an [out] variable. nullptr is only for srv_pass_corrupt_tables and it should be check with macro that will assert if the srv_pass_corrupt_tables != ON. Will update the patch with this assert.

Comment (robgolebiowski) [reply storage/innobase/dict/dict0dict.cc:7504]: OK, I have confused myself for a moment - both for decryption errors and for srv_pass_corrupt_tables the page can be nullptr.

**Pull request merged**
